### ISSUE TITLE
In TLS1.3, key_share make debug is hard.

### ISSUE THREAD

**Issue by fengmao31** [openssl/openssl#19169]
I test two websockets program to buid a TLS handshark. But after the first handshark, they use the key_share to recover the before TLS session ticket. And I cannot catch the certificates to check my program.Then I move the server to another computer. I find it still use the before key_share. 
My client is in the VMware. And the server first is in the VMware in the same computer. Then I move it to another pysical computer not a VMware in the same LAN. It uses the old key_share.
I make some certificates to replace the certs in the server and change the verify mode from Null to Peer. It still use the old session.
It is too dangerous and crazy. How to clean the cache of the old session.

**Comment by mattcaswell** [comment]
Your question is a little unclear, but it sounds like you may be getting confused about what the purpose of the key_share extension is. This extension will exist every time a TLSv1.3 capable OpenSSL client attempts to connect to a server, regardless of whether it is attempting to resume an old session or not. It is used to derive a connection specific handshake/master secret. There may also be a PSK from a previous session (sent in the psk extension) which is combined with the key_share value to derive the connection specific handshake/master secret - but this is optional.

Note the key_share value should be different for each connection attempt - so you should never see an old value repeated from an earlier connection attempt.

**Comment by fengmao31** [comment]
I want to clear the session cache. because I am going to debug the websockets server and client. but every time it use the key_share to recover the old handshark.
even I change the server certificate, client the verify mode, moving the server to another computer, it  still use the  key_share to recover the handshark session.
![image](https://user-images.githubusercontent.com/19343842/189095532-3087e961-3c12-4af7-9f35-d3e6a024690b.png)

I must stop TLS 1.3 mothed to get the certificate by using TLS 1.2 mothed.

![image](https://user-images.githubusercontent.com/19343842/189095424-9261c11c-b7ea-4ccc-94ba-fc2c8cd86336.png)

**Comment by mattcaswell** [comment]
> but every time it use the key_share to recover the old handshark.

key_share is *always* sent by an OpenSSL TLSv1.3 capable client. It is always different. It is not used to resume an old session.

**Comment by fengmao31** [comment]
but why they don‘t exchange the certificates in TLS1.3. I want to catch the certificate.

**Comment by mattcaswell** [comment]
> but why they don‘t exchange the certificates in TLS1.3. I want to catch the certificate.

All the handshake after the initial ClientHello/ServerHello is encrypted in TLSv1.3 and will appear identical to Application Data (and will be reported as such by Wireshark). This includes any Certificate message. In all probability a certificate is being sent - but you just can't see it in Wireshark. You would need to decrypt the handshake in Wireshark to do this - not sure if it supports it or not (I've never attempted it).

**Comment by hlandau** [comment]
Yes, wireshark supports this. You will need to dump a keylog using `SSL_CTX_get_keylog_callback`, writing the provided data to a file. You can then set the path to the file in Wireshark settings. Alternatively you can embed the keylog file directly into a `.pcapng` file using the editcap tool: `editcap --inject-secrets tls,$KEYLOG_FILE in.pcapng out.pcapng`. Wireshark can then use the keylog automatically.

**Comment by fengmao31** [comment]
I will try to check it. Thank you.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the end of the 3.4 dev cycle barring further input
